bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py

```diff
@@ -45,23 +45,19 @@ def RDataFrameAsNumpy(df, columns=None, exclude=None):
     Parameters:
         columns: If None return all branches as columns, otherwise specify names in iterable.
         exclude: Exclude branches from selection.
+        lazy: Determines whether this action is instant (False, default) or lazy (True).
 
     Returns:
-        dict: Dict with column names as keys and 1D numpy arrays with content as values
+        dict or AsNumpyResult: if instant (default), dict with column names as keys and
+            1D numpy arrays with content as values; if lazy, AsNumpyResult containing
+            the result pointers obtained from the Take actions.
     """
     # Sanitize input arguments
     if isinstance(columns, str):
         raise TypeError("The columns argument requires a list of strings")
     if isinstance(exclude, str):
         raise TypeError("The exclude argument requires a list of strings")
 
-    # Import numpy and numpy.array derived class lazily
-    try:
-        import numpy
-        from ROOT.pythonization._rdf_utils import ndarray
-    except:
-        raise ImportError("Failed to import numpy during call of RDataFrame.AsNumpy.")
-
```

Comment:
I see your point, but this would cause a repetition of this code that I don't like. Note that when the action is instant (default), the import error would still happen at `AsNumpy` call time. But if you declare it lazy, everything happens lazily (even the error). I don't think that's too unfair.

Reply:
> this would cause a repetition of this code that I don't like

i was not suggesting to repeat the check but to only have it here, where `AsNumpy` is called, is that not enough?

Reply:
I still think that laziness also means complaining about missing stuff when it's actually required, but ok, I'll leave the try except with numpy here and the two imports unprotected down in GetValue.

Reply:
> I still think that laziness also means complaining about missing stuff when it's actually required

Yes, absolutely, but we _know_ numpy will be required the moment users call `AsNumpy`, right?

Reply:
If we are strict that's not completely true, because it could happen that the result of `AsNumpy` is never claimed. In that case, numpy is actually never needed.

Reply:
ah come on now :smile: let's code for the common usecase

Reply:
If I construct an RDF with a wrong tree name, that's fine until I do something with that RDF, then I see the error. Is that conceptually different than the case here?

Reply:
I don't know where you are going here, don't you agree that erroring out early and at the point of call is more helpful and simpler for users than erroring out from a different place and possibly much later than when the problem was introduced?

(there are considerations that make it trickier for RDF's constructor than for the case here, but sure, in general it would be nice to tell users "hey, you are saying that you want to process this tree and file, but they don't exist" at the point they construct the RDF)

Reply:
I'll move it, it's fine. I don't have the same perception as you that erroring out at the moment when NumPy is actually needed is bad, since we are working with futures and laziness. But I agree that the early message can help users.